### PR TITLE
Include an Audit Script for the Colors Used in SVG Images

### PR DESCRIPTION
**Pull request by porada**
**…and Update the Remaining Images**

This pull request introduces an auxiliary script that detects the use of non-standard color values in the SVG images found in the repository.

This script has been tested with the updates performed in #36510 (as a part of #36097). Some images were left out from that PR due to quirky color substitution, so they have been updated in this PR instead (along with the tweaks to the script). Our retired plan icons have been removed instead of receiving the facelift.

### Testing instructions

* Make sure running following command doesn’t throws errors:

    ```sh
    node ./bin/audit-svg-colors.js
    ```

* Ensure the updated images show up in Calypso.

---

☝️ @sixhours @davewhitley @Aurorum

### CODE REVIEW

**Comment by porada** [comment]
This PR is considered to be still in progress until #36510 confirms it works as intended.

Here’s the example output for now:

```
$ ./bin/audits-svg-colors.js

Found 192 SVG images in this repository that use non-standard color values:

client/assets/images/illustrations/custom-domain-blogger.svg
#ccced0 → #c3c4c7 (Gray 10)
#969ca1 → #8e9196 (Gray 30)
#46799a → #187aa2 (WordPress Blue 40)

client/assets/images/illustrations/custom-domain.svg
#ccced0 → #c3c4c7 (Gray 10)
#969ca1 → #8e9196 (Gray 30)
#6f93ad → #3895ba (WordPress Blue 30)
#46799a → #187aa2 (WordPress Blue 40)
#204a69 → #004e6e (WordPress Blue 60)

…
```
```
$ ./bin/audits-svg-colors.js --list

Found 135 non-standard color values used in the SVG images in this repository:

#d0f4e7 → #e4f2ed (Celadon 0)
#7bdebf → #63d6b6 (Celadon 10)
#44d0a3 → #2ebd99 (Celadon 20)
#61b87c → #2ebd99 (Celadon 20)
#f6f6f6 → #f6f7f7 (Gray 0)
#f4f8fa → #f6f7f7 (Gray 0)
#f1f5f8 → #f6f7f7 (Gray 0)
#f2f5f7 → #f6f7f7 (Gray 0)
#f3f3f3 → #f6f7f7 (Gray 0)
#fbfcfd → #f6f7f7 (Gray 0)
#f4f4f4 → #f6f7f7 (Gray 0)
#fefefe → #f6f7f7 (Gray 0)
…
```

Meanwhile, if anybody has any suggestions or comments on the script, please let me know. 🙂

**Comment by simison** [comment]
I wonder if it would make sense to expose this as part of the [Color studio](https://github.com/Automattic/color-studio/)?

If you add the file and this in color studio’s `package.json`
```js
  "bin": {
    "audit-svg-colors": "bin/audit-svg-colors.js"
  }
```

... you could run `npx audit-svg-colors` and add some way to pass excluded list to that script from Calypso, or from Jetpack or elsewhere?

**Comment by simison** [comment]
Would it be useful to add this to CI and fail if someone tries to add offending SVGs in PRs?

**Comment by simison** [comment]
An interesting future iteration of this could also be a config for [SVGLint](https://github.com/birjolaxew/svglint) or as an [Eslint plugin](https://eslint.org/docs/developer-guide/working-with-plugins).

**Comment by porada** [comment]
Thanks for the feedback, @simison!

### The package

I pitch Color Studio the `npm` package as dependency-free and I’d rather keep it that way unless we have a really good reason not to. The audit script has one major dependency so far, and if we were to add more features to it, that list would probably grow.

The alternative way would be to extract this as separate package (e.g. `@automattic/color-studio-audit`) and let it live its own life.

### CI and linting

I’d love if we could automatically run this on CI or as a pre-commit check. Before that happens, though, the script would require a few additional features. One of those is supplying paths to check as an input, so that we could lint only the files have been modified in a current branch.

Apart from time, I’d say that the constraint of having a single file is what kept me from including more features (with `--fix` being on the top of my list). The file grows in size with every single addition and is already on the edge of being readable or not.

That’s probably another reason to move it into a separate self-contained package. And if we had that, integrating with other linters would probably be much easier.

### 20/80

Having said the above, I’m confident that this script in its current form is the 20% of work that will help us phase out 80% or more inconsistent cases that we would have never caught manually.

**Comment by simison** [comment]
> Having  said the above, I’m confident that this script in its current form is  the 20% of work that will help us phase out 80% or more inconsistent  cases that we would have never caught manually.

Absolutely! My comments are just food for thought. :-)

File: bin/audit-svg-colors.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+/**
+ * The script looks for all colors used in the SVG images found in the repository
+ * and suggests Color Studio replacements for all non-standard values it finds.
+ *
+ * List all non-standard color values by image path:
+ * $ ./bin/audits-svg-colors.js
+ *
+ * List all non-standard color values in aggregate:
+ * $ ./bin/audits-svg-colors.js --list
+ */
```
Comment (Aurorum) [inline bin/audit-svg-colors.js]: Is this meant to be `./bin/audit-svg-colors.js` based on the file name? I got a Permission Denied error when trying it, but whilst running this, I got a "file not found" error.

**Comment by porada** [comment]
Fixed typos and rebased.

**Comment by porada** [comment]
Removed the support for `--list` as some images may use different color subsets (defined in the script).

File: bin/audit-svg-colors.js [hunk 2]
```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env node
+
+/**
+ * The script looks for all colors used in the SVG images found in the repository
+ * and suggests Color Studio replacements for all non-standard values it finds.
+ *
+ * Please note that the suggestions are automattic and therefore not perfect
```
Comment (sirreal) [inline bin/audit-svg-colors.js]: Gets me every time 😂 

```suggestion
 * Please note that the suggestions are automatic and therefore not perfect
```

File: bin/audit-svg-colors.js [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+	// Gutenberg images
+	/^static\/images\/illustrations\/gutenberg/,
+
+	// Plan icons
+	/^static\/images\/plans\//,
+];
+
+// The regular expression used to identify color values
+const SVG_VALUE_EXPRESSION = /(?:fill|flood-color|lighting-color|stop-color|stroke)="([a-z0-9#]*?)"/gi;
```
Comment (sirreal) [inline bin/audit-svg-colors.js]: If this is intended to target hex values, we could tighten it up a bit.

There are more valid values than hex that we'll find here. That's probably acceptable for a tool like this.

**Review by sirreal — COMMENTED** [review]
I looked this over and it seems sane. I did not test it.

I've added @Automattic/team-calypso for review. They're probably the team best suited for reviewing tooling changes and additions like this.

If there's value here, it makes sense to add tools like this.

I notice that there are some visual changes in this PR. That should probably be mentioned in the description, preferably with before/after images of anything significant.

File: bin/audit-svg-colors.js (hunk 2 above)
Comment (porada) [reply bin/audit-svg-colors.js]: Yeah…

File: bin/audit-svg-colors.js (hunk 3 above)
Comment (porada) [reply bin/audit-svg-colors.js]: This also captures explicit color names like `white`, `black`, etc.

I’m leaving out `rgba()` and other similar formats as I’ve never seen them using within a SVG file.

**Comment by porada** [comment]
@sirreal, thanks for chiming in!

> I notice that there are some visual changes in this PR. That should probably be mentioned in the description, preferably with before/after images of anything significant.

Those are just minor updates to the color values used in those files and most will be almost invisible in testing. Not sure that listing them out individually will help—especially when we have cases of the updates not being present right away (which might be caused by CDN cache).

What testing strategy do you think would be best for those visual changes?

**Comment by porada** [comment]
Speaking of added value:

> If there's value here, it makes sense to add tools like this.

We executed most of #36510 with the script’s help. The visual changes here are the edge cases we caught during that process.

**Comment by sirreal** [comment]
> Those are just minor updates to the color values used in those files and most will be almost invisible in testing.

👌 

> What testing strategy do you think would be best for those visual changes?

I don't believe we've ever had a good story for visual regression testing. I'm not sure what the reasons are. @bsessions85 or @Stojdza might have more information.

**Comment by porada** [comment]
@Automattic/team-calypso, can we get back to the discussion about merging this? I’m cooking up another round of our color palette updates and would benefit from being able to use this script.

**Comment by porada** [comment]
@blowery, rebased as discussed and ready for your review.

**Comment by blowery** [comment]
@porada package-lock looks busted. Try running `npm i` again from the root and see if it gets updated?

**Comment by porada** [comment]
I initially ignored the error, because I assumed it was an issue with `@wordpress/scripts` which we’ve also been experiencing in other projects. But I just pushed a fix—let’s see if the tests succeed now.

**Review by blowery — APPROVED** [review]
Nice, looks good now!

**Comment by blowery** [comment]
> because I assumed it was an issue with @wordpress/scripts which we’ve also been experiencing in other projects.

Could you elaborate on that more? Was this reported already?

**Comment by porada** [comment]
Thanks for the review, @blowery!

**Comment by porada** [comment]
> Could you elaborate on that more? Was this reported already?

Whenever we would touch dependencies, a simple `npm install` would end up with the following error:

```
[format:js] > wp-scripts format-js
[format:js] 
[format:js] Incompatible version of Prettier was found in your project
[format:js] You need to install the 'wp-prettier' package to get code formatting compliant with the WordPress coding standards.
```

In most cases, removing `package-lock.json` and rewriting it from scratch would help (but I think that’s missing the point of having one in the repo in the first place). I was pretty sure one of my MarTech teammates had investigated and reported it, but looks like that didn’t happen. I’m planning on reporting it whenever it pops up again for us.

**Pull request merged**